components/cluster-coordinator/org.wso2.carbon.cluster.coordinator/org.wso2.carbon.cluster.coordinator.rdbms/src/main/java/org/wso2/carbon/cluster/coordinator/rdbms/RDBMSCoordinationStrategy.java

```diff
@@ -210,20 +222,60 @@ public void joinGroup(Map<String, Object> propertiesMap) {
             long currentTimeMillis = System.currentTimeMillis();
             long heartbeatAge = currentTimeMillis - lastHeartBeat;
-            isNodeExist = (heartbeatAge < heartbeatMaxRetry);
+            isNodeExist = (heartbeatAge < heartbeatMaxRetryInterval);
         }
 
         if (!isNodeExist) {
-            CoordinatorElectionTask coordinatorElectionTask = new CoordinatorElectionTask(localNodeId,
-                    localGroupId, propertiesMap);
-            this.threadExecutor.scheduleAtFixedRate(coordinatorElectionTask, 0, heartBeatInterval,
-                    TimeUnit.MILLISECONDS);
+            isCoordinatorTasksRunning = true;
+            this.threadExecutor.execute(new HeartBeatExecutionTask(propertiesMap));
         } else {
             throw new ClusterCoordinationException("Node with ID " + localNodeId + " in group " + localGroupId +
                     " already exists.");
         }
     }
 
+    /**
+     * This class will schedule and execute coordination tasks
+     */
+    public class HeartBeatExecutionTask implements Runnable {
+        private CoordinatorElectionTask coordinatorElectionTask;
+        private long lastHeartbeatFinishedTime;
+
+        public HeartBeatExecutionTask(Map<String, Object> propertiesMap) {
+            coordinatorElectionTask = new CoordinatorElectionTask(localNodeId, localGroupId, propertiesMap);
+        }
+
+        @Override
+        public void run() {
+            while (isCoordinatorTasksRunning) {
+                try {
+                    long currentHeartbeatStartedTime = System.currentTimeMillis();
+                    coordinatorElectionTask.runCoordinationElectionTask(currentHeartbeatStartedTime);
+                    long taskEndedTime = System.currentTimeMillis();
+                    if (lastHeartbeatFinishedTime != 0 &&
+                            ((taskEndedTime - (lastHeartbeatFinishedTime + heartBeatInterval))
+                                    >= heartbeatWarningMargin)) {
+                        log.warn("The heartBeatInterval is in " + heartBeatInterval +
+                                " millis with a retry count of " + heartbeatMaxRetry + ". " +
+                                "But current heartbeat has happened after " +
+                                (currentHeartbeatStartedTime - lastHeartbeatFinishedTime) +
+                                " millis from the last heartbeat, and took " +
+                                (taskEndedTime - currentHeartbeatStartedTime) +
+                                " millis to run CoordinationElection on the database at " +
+                                currentHeartbeatStartedTime +
+                                ". Please increase the heartBeat interval or the retry count.");
+                    }
+                    lastHeartbeatFinishedTime = currentHeartbeatStartedTime;
+                    if (lastHeartbeatFinishedTime + heartBeatInterval - taskEndedTime > 5) {
```

Comment:
what is this 5 refers  to ?

Reply:
This is to avoid sleeping for very short periods.